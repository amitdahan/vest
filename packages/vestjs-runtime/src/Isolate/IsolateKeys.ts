import { assign } from 'lodash';

export enum IsolateKeys {
  Type = '$type',
  Keys = 'keys',
  Key = 'key',
  Parent = 'parent',
  Data = 'data',
  AllowReorder = 'allowReorder',
  Status = 'status',
  AbortController = 'abortController',
  Children = 'children',
}

export enum MinifiedKeys {
  Type = '$',
  Keys = 'Ks',
  Key = 'ky',
  Parent = 'P',
  Data = 'D',
  AllowReorder = 'AR',
  Status = 'S',
  Children = 'C',
}

export const KeyToMinified = {
  [IsolateKeys.Type]: MinifiedKeys.Type,
  [IsolateKeys.Keys]: MinifiedKeys.Keys,
  [IsolateKeys.Parent]: MinifiedKeys.Parent,
  [IsolateKeys.Data]: MinifiedKeys.Data,
  [IsolateKeys.Key]: MinifiedKeys.Key,
  [IsolateKeys.AllowReorder]: MinifiedKeys.AllowReorder,
  [IsolateKeys.Status]: MinifiedKeys.Status,
  [IsolateKeys.Children]: MinifiedKeys.Children,
};

// This const is an object that looks like this:
// {
//   '$': '$type',
//   'K': 'keys',
//   'P': 'parent',
//   ...
// }
export const MinifiedToKey = invertKeyMap(KeyToMinified);

export function invertKeyMap(miniMap: Record<string, string> = {}) {
  return Object.entries(miniMap).reduce(
    (acc, [key, minified]) =>
      assign(acc, {
        [minified]: key,
      }),
    {} as Record<string, string>
  );
}

export const ExcludedFromDump = [
  IsolateKeys.AbortController,
  IsolateKeys.Parent,
  IsolateKeys.Keys,
  IsolateKeys.Children,
  IsolateKeys.Data,
];
