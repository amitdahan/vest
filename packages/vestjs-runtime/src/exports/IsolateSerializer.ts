import { ErrorStrings } from 'ErrorStrings';
import {
  Maybe,
  Nullable,
  hasOwnProperty,
  invariant,
  isEmpty,
  isNotNullish,
  isNullish,
  isStringValue,
  text,
} from 'vest-utils';

import { TIsolate } from 'Isolate';
import {
  ExcludedFromDump,
  IsolateKeys,
  KeyToMinified,
  MinifiedKeys,
  MinifiedToKey,
  invertKeyMap,
} from 'IsolateKeys';
import { IsolateMutator } from 'IsolateMutator';

export class IsolateSerializer {
  // eslint-disable-next-line max-statements, complexity
  static deserialize(
    node: Record<string, any> | TIsolate | string,
    miniMaps: Maybe<MiniMaps>
  ): TIsolate {
    // the  assumption is that the tree is built correctly,
    // but the children are missing the parent property to
    // avoid circular references during serialization.
    // in the same way, the parents are missing the `keys` property
    // to avoid circular references during serialization.
    // we need to rebuild the tree and add back the parent property to the children
    // and the keys property to the parents.
    const inverseMinimap = invertKeyMap(miniMaps?.keys?.data);

    // Validate the root object
    const root = isStringValue(node)
      ? JSON.parse(node)
      : ({ ...node } as TIsolate);

    IsolateSerializer.validateIsolate(root);

    const queue = [root];

    // Iterate over the queue until it's empty
    while (queue.length) {
      // Get the next item from the queue
      const current = queue.shift();

      // Get the children of the current item
      const children = IsolateSerializer.expandChildren(current);

      // Iterate over the minified keys
      for (const key in MinifiedToKey) {
        // Get the value for the current key
        const value = current[key];

        // If the value is not null or undefined
        if (isNotNullish(value)) {
          // Get the key to use
          const keyToUse = MinifiedToKey[key];

          // If the key is data, then we may need to transform the keys
          if (keyToUse === IsolateKeys.Data) {
            // Transform the keys
            current[keyToUse] = transformKeys(value, inverseMinimap);
          } else {
            // Otherwise, just set the key
            current[keyToUse] = value;
          }

          // Remove the old key
          delete current[key];
        }
      }

      // If there are no children, nothing to do.
      if (!children) {
        continue;
      }

      // Copy the children and set their parent to the current node.
      current.children = children.map(child => {
        const nextChild = { ...child };

        IsolateMutator.setParent(nextChild, current);
        queue.push(nextChild);

        // If the child has a key, add it to the parent's keys.
        const key = nextChild.key;

        if (key) {
          current.keys = current.keys ?? {};
          current.keys[key] = nextChild;
        }

        return nextChild;
      });
    }

    return root as TIsolate;
  }

  static serialize(
    isolate: Nullable<TIsolate>,
    miniMaps: Maybe<MiniMaps>
  ): string {
    if (isNullish(isolate)) {
      return '';
    }

    return JSON.stringify(transformIsolate(isolate, miniMaps));
  }

  static expandChildren(node: Record<string, any>): Nullable<TIsolate[]> {
    return node[MinifiedKeys.Children]
      ? [...node[MinifiedKeys.Children]]
      : null;
  }

  static validateIsolate(node: Record<string, any> | TIsolate): void {
    invariant(
      hasOwnProperty(node, IsolateKeys.Type) ||
        hasOwnProperty(node, KeyToMinified[IsolateKeys.Type]),
      text(ErrorStrings.INVALID_ISOLATE_CANNOT_PARSE)
    );
  }
}

// eslint-disable-next-line max-statements, complexity
function transformIsolate(
  isolate: TIsolate,
  miniMaps: Maybe<MiniMaps>
): Record<string, any> {
  const next: Record<string, any> = {};

  if (isolate.children) {
    next[MinifiedKeys.Children] = isolate.children.map(isolate =>
      transformIsolate(isolate, miniMaps)
    );
  }

  if (!isEmpty(isolate.data)) {
    next[MinifiedKeys.Data] = transformKeys(isolate.data, miniMaps?.keys?.data);
  }

  for (const key in isolate) {
    // Skip keys that should be excluded from the dump.
    // While we're excluding children from the dump, they'll actually remain there
    // due to the fact that we've already transformed them recursively beforehand
    // thus renaming them to the minified key.
    if (isKeyExcluededFromDump(key)) {
      continue;
    }
    const value = isolate[key as keyof TIsolate];

    if (isNullish(value)) {
      continue;
    }

    const keyToUse = minifyKey(key);
    next[keyToUse] = minifyValueByKey(key, value, miniMaps);
  }

  return next;
}

function isKeyExcluededFromDump(key: string): boolean {
  return ExcludedFromDump.includes(key as IsolateKeys);
}

function minifyKey(key: string): string {
  return KeyToMinified[key as keyof typeof KeyToMinified] ?? key;
}

function minifyValueByKey(
  key: string,
  value: any,
  miniMaps: Maybe<MiniMaps>
): any {
  if (isNullish(value)) {
    return value;
  }

  const keyMap = miniMaps?.values?.[key as keyof MiniMaps['values']];

  return keyMap ? keyMap[value] ?? value : value;
}

function transformKeys(
  data: Record<string, any>,
  keyMap: Maybe<MiniMap>
): Record<string, any> {
  const next: Record<string, any> = {};

  // Loop over each key in the data.
  for (const key in data) {
    // Find the key to use for the next object.
    // If there is no key map, use the original key.
    // If there is a key map, use the key map entry for the current key.
    const keyToUse = (keyMap ? keyMap[key] : key) ?? key;

    // Add the key and value to the next object.
    next[keyToUse] = data[key];
  }

  // Return the next object.
  return next;
}

type MiniMap = Record<string, string>;

type MiniMaps = Partial<{
  keys: Partial<{
    [IsolateKeys.Data]: MiniMap;
  }>;
  values: Partial<{
    [IsolateKeys.Status]: MiniMap;
    [IsolateKeys.Type]: MiniMap;
  }>;
}>;
