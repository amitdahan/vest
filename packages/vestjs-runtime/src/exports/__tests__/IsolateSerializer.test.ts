import { Maybe } from 'vest-utils';

import { Isolate, TIsolate } from 'Isolate';
import { MinifiedKeys } from 'IsolateKeys';
import { IsolateSerializer } from 'IsolateSerializer';
import { VestRuntime } from 'vestjs-runtime';

describe('IsolateSerializer', () => {
  describe('serialize', () => {
    it('Should produce serialized dump', () => {
      const { serialized } = createSerialized();

      expect(serialized).toMatchSnapshot();
    });
  });

  describe('deserialize', () => {
    it('Should fully inflate the tree', () => {
      const { root, serialized } = createSerialized();

      const inflated = IsolateSerializer.deserialize(serialized);

      expect(inflated).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "children": [
            {
              "$type": "UChild_1",
              "data": {
                "some_data": true,
              },
              "parent": [Circular],
            },
            {
              "$type": "UChild_2",
              "parent": [Circular],
            },
            {
              "$type": "UChild_3",
              "parent": [Circular],
            },
          ],
          "data": {
            "some_data": true,
          },
        }
      `);
    });

    test('When data is empty, should not add data property', () => {
      const { serialized } = createSerialized();
      const inflated = IsolateSerializer.deserialize(serialized);

      expect(inflated?.children?.[0]).toHaveProperty('data');
      expect(inflated?.children?.[1]).not.toHaveProperty('data');
      expect(inflated?.children?.[2]).not.toHaveProperty('data');
    });
  });

  describe('Custom Data Serialization', () => {
    it('Should serialize data with custom keys', () => {
      const { serialized } = createSerialized({
        some_data: 'sd',
      });

      const parsed = JSON.parse(serialized);
      expect(parsed[MinifiedKeys.Data]).toHaveProperty('sd');
      expect(serialized).toMatchInlineSnapshot(
        `"{"C":[{"D":{"sd":true},"$":"UChild_1"},{"$":"UChild_2"},{"$":"UChild_3"}],"D":{"sd":true},"$":"URoot"}"`
      );
    });

    it('Should inflate with correct keys', () => {
      const { serialized } = createSerialized({
        some_data: 'sd',
      });

      const inflated = IsolateSerializer.deserialize(serialized, {
        some_data: 'sd',
      });

      expect(inflated.data.some_data).toBe(true);
      expect(inflated).not.toHaveProperty('sd');
      expect(inflated).toEqual(
        IsolateSerializer.deserialize(createSerialized().serialized)
      );
      expect(inflated).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "children": [
            {
              "$type": "UChild_1",
              "data": {
                "some_data": true,
              },
              "parent": [Circular],
            },
            {
              "$type": "UChild_2",
              "parent": [Circular],
            },
            {
              "$type": "UChild_3",
              "parent": [Circular],
            },
          ],
          "data": {
            "some_data": true,
          },
        }
      `);
    });
  });
});

function withRunTime<T>(fn: CB<T>) {
  return VestRuntime.Run(VestRuntime.createRef({}), () => {
    return fn();
  });
}

function createSerialized(miniMap: Maybe<Record<string, string>>) {
  let serialized: string, root: TIsolate;

  withRunTime(() => {
    root = Isolate.create(
      'URoot',
      () => {
        Isolate.create('UChild_1', () => {}, { some_data: true });
        Isolate.create('UChild_2', () => {});
        Isolate.create('UChild_3', () => {});
      },
      {
        some_data: true,
      }
    );

    serialized = IsolateSerializer.serialize(root, miniMap);
  });

  // @ts-ignore
  return { root, serialized };
}
