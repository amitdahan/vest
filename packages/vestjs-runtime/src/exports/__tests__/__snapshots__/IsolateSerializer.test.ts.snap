// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsolateSerializer serialize Should produce serialized dump 1`] = `"{"C":[{"D":{"some_data":true},"$":"UChild_1"},{"$":"UChild_2"},{"$":"UChild_3"}],"D":{"some_data":true},"$":"URoot"}"`;
